src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,176 @@
+        }
+        if (index != 0 && index != size) {
+            Node<T> currentNode = findNode(index);
+            Node<T> addNode = new Node<>(null, value, currentNode);
+            addNode.previous = currentNode.previous;
+            currentNode.previous.next = addNode;
+            currentNode.previous = addNode;
+            size++;
+            return;
+        }
+        if (index == 0) {
+            Node<T> addNode = new Node<>(null, value, head);
+            head.previous = addNode;
+            head = addNode;
+            size++;
+            return;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return findNode(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> currentNode = findNode(index);
+        T tempItem = currentNode.item;
+        currentNode.item = value;
+        return tempItem;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> currentNode = findNode(index);
+        unlinkNode(currentNode);
+        return currentNode.item;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> searchNode = head;
+        while (searchNode != null) {
+            if (searchNode.item == object
+                    || (searchNode.item != null && searchNode.item.equals(object))) {
+                unlinkNode(searchNode);
+                return true;
+            }
+            searchNode = searchNode.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void unlinkNode(Node node) {
+        if (node.previous == null && node.next != null) {
+            node.next.previous = null;
+            head = node.next;
+            node.next = null;
+            node.previous = null;
+            size--;
+            return;
+        }
+        if (node.next == null && node.previous != null) {
+            node.previous.next = null;
+            tail = node.previous;
+            node.next = null;
+            node.previous = null;
+            size--;
+            return;
+        }
+        if (node.previous == null && node.next == null) {
+            node.next = null;
+            node.previous = null;
+            head = null;
+            tail = null;
+            size = 0;
+            return;
+        }
+        node.previous.next = node.next;
+        node.next.previous = node.previous;
+        node.next = null;
+        node.previous = null;
+        size--;
+    }
+
+    private Node<T> findNode(int index) {
+        checkIndexValidity(index);
+        if ((index <= size / 2) && head != null) {
```

Comment:
do we need this check `head != null` ?

Reply:
thanks. Next if has checking up for head ! = null. It's not necessary